ambry-store/src/test/java/com.github.ambry.store/CompactionManagerTest.java

```diff
@@ -150,14 +175,110 @@ public void testDifferentMessageRetentionDays() throws StoreException, Interrupt
+      store.validLogSegments = i == 0 ? null : generateRandomStrings(i);
+      stores.add(store);
+    }
+    compactionManager = new CompactionManager(MOUNT_PATH, config, stores, time);
+    compactionManager.enable();
+    assertTrue("Compaction calls did not come within the expected time",
+        compactCallsCountdown.await(1, TimeUnit.SECONDS));
+    for (int i = 0; i < numStores; i++) {
+      MockBlobStore store = (MockBlobStore) stores.get(i);
+      if (store.callOrderException != null) {
+        throw store.callOrderException;
+      }
+      if (i > 0) {
+        assertTrue("Compact was not called", store.compactCalled);
+      } else {
+        // should not call for i == 0 because there are no compaction details.
+        assertFalse("Compact should not have been called", store.compactCalled);
+      }
+    }
+    compactionManager.disable();
+    compactionManager.awaitTermination();
+  }
+
+  /**
+   * Tests that compaction proceeds on all non misbehaving stores even in the presence of some misbehaving stores
+   * (not started/throwing exceptions).
+   * @throws Exception
+   */
+  @Test
+  public void testCompactionWithMisbehavingStores() throws Exception {
+    int numStores = 5;
+    List<BlobStore> stores = new ArrayList<>();
+    // one store that isn't started isn't going to get compact calls.
+    // another store that throws an exception on resumeCompaction() isn't going to get a compact call.
+    CountDownLatch compactCallsCountdown = new CountDownLatch(numStores - 2);
+    properties.setProperty("store.enable.compaction", Boolean.toString(true));
+    config = new StoreConfig(new VerifiableProperties(properties));
+    MetricRegistry metricRegistry = new MetricRegistry();
+    StorageManagerMetrics metrics = new StorageManagerMetrics(metricRegistry);
+    for (int i = 0; i < numStores; i++) {
+      MockBlobStore store = new MockBlobStore(config, metrics, time, CAPACITY_IN_BYTES, DEFAULT_USED_CAPACITY_IN_BYTES,
+          compactCallsCountdown);
+      store.validLogSegments = generateRandomStrings(2);
+      if (i == 0) {
+        // one store should not be started
+        store.started = false;
+      } else if (i == 1) {
+        // one store should throw on resumeCompaction()
+        store.exceptionToThrowOnResume = new RuntimeException("Misbehaving store");
+      } else if (i == 2) {
+        // one store should throw on compact()
+        store.exceptionToThrowOnCompact = new RuntimeException("Misbehaving store");
+      }
+      stores.add(store);
+    }
+    compactionManager = new CompactionManager(MOUNT_PATH, config, stores, time);
+    compactionManager.enable();
+    assertTrue("Compaction calls did not come within the expected time",
+        compactCallsCountdown.await(1, TimeUnit.SECONDS));
+    for (int i = 0; i < numStores; i++) {
+      MockBlobStore store = (MockBlobStore) stores.get(i);
+      if (store.callOrderException != null) {
+        throw store.callOrderException;
```

Comment:
`assertNull` does not give me any information on what the exception was - if just failing was the desired behavior, could have just used a `boolean`. In this case, I'm interested in what the actual problem was which is why failing the test with the actual exception is useful.